### DISCUSSION TITLE
Redirect after refreshUser() in UserProvider invalidated my token

### DISCUSSION THREAD

**Discussion by JimPanse2k19** [symfony/symfony#43026] [Q&A]
Hello,

i desperately try to get my user redirected to an external page after my refreshToken() method in my UserProvider invalidated my user session by either throwing an UsernameNotFoundException or just returning a different user.

Nothing of this seems to work since no event is thrown after the token in the token storage was set to null (and in fact my user was logged out with this). So i would expect at least some kind of logout event here (?)

I am bound to version 5.2, so no chance to update now

Here is the background:

I have some external SSO solution where my user logs in. The token for this gets saved in the session of symfony. Once logged in (realized by an authenticator exchanging the auth code by an access token) symfony calls on every request the refreshToken() method in my user provider to check if the user is still valid. 

Here is my logic to check if the symfony token still complies to the expiration time of the access token from the session etc.

If not, the user changed and/or should be not available anymore ...

What i tried:

1. Throw an UsernameNotFoundException -> nothing happens, just the user is logged out from symfony, no chance to redirect or to react on events somewhere

2. Manipulate the user to trigger the DeauthenticatedEvent ... it works, but different than expected, returning a redirectresponse here does also have no effect. So, it has no use to me too.

How can i Redirect the user after the token is gone? What do i miss?

Best regards

**Comment by jfsenechal** [comment]
Hello

did you look here

https://symfony.com/doc/current/security/user_provider.html#custom-user-provider

or maybe that

https://symfony.com/doc/current/security/user_checkers.html

Work on the event Kernel request ?

Symfony\Component\HttpKernel\EventListener\SessionListener::onKernelRequest()

**Reply by JimPanse2k19** [reply]
Like i wrote, the user providers refreshUser() method does just set the token to null ... but do not fire any other events i could react on afterwards ... i wonder if this isn't a usual use case to "react" on changes like that.
Does the UserChecker fire any events?

**Comment by jfsenechal** [comment]
as you indicate there is no event on refreshToken

So you have to work on an existing event you can do it with the command below

```bash
bin/console make:subscriber
```
we offer you the list of events I think kernel.request is suitable

```php

use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpKernel\Event\RequestEvent;
use Symfony\Component\Routing\RouterInterface;
use Symfony\Component\Security\Core\Security;

class UserNotMoreLoginSubscriber implements EventSubscriberInterface
{
    private Security $security;
    private RouterInterface $router;

    public function __construct(Security $security, RouterInterface $router)
    {
        $this->security = $security;
        $this->router = $router;
    }

    public function onKernelRequest(RequestEvent $event)
    {
        //no request ajax
        if ($event->isMainRequest()) {
            //check if user is is still logged in
            if ($this->security->getToken() === null) {
                return new RedirectResponse($this->router->generate('app_x'));
            }
        }
    }

    public static function getSubscribedEvents()
    {
        return [
            'kernel.request' => 'onKernelRequest',
        ];
    }
}
```

**Reply by JimPanse2k19** [reply]
Thanks a lot for your help. It seems that this is the right direction so far. 

But it looks like checking for an empty token storage will not do the trick. The token storage is not empty anymore on my kernel event listener. The AnonymousAuthenticationListener seems to set an anonymous token immediately after the ContextListener set it to null.
I guess, something like checking for grants will come into game now?

**Reply by jfsenechal** [reply]
Yes maybe with   $this->security->isGranted()

**Reply by JimPanse2k19** [reply]
I don't get it ... the event seems not to have a Response attached so the redirect will not happen ... isn't setResponse() the correct way to set it instead of return?

**Comment by jfsenechal** [comment]
possible I did not try the code I gave you

**Reply by jfsenechal** [reply]
the correct event is kernel.response

```php
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpFoundation\RedirectResponse;
use Symfony\Component\HttpKernel\Event\ResponseEvent;

class ExceptionSubscriber implements EventSubscriberInterface
{
    public function onKernelResponse(ResponseEvent $event)
    {
        if (!$event->isMainRequest()) {
            // don't do anything if it's not the main request
            return;
        }
        if ($condition) {
            $event->setResponse(new RedirectResponse('/lolo'));
        }
    }

    public static function getSubscribedEvents()
    {
        return [
            'kernel.response' => 'onKernelResponse',
        ];
    }
}
```
